src/drivers/stm32/adc/adc.cpp

```diff
@@ -455,30 +428,21 @@ ADC::update_system_power(hrt_abstime now)
 #ifdef BOARD_ADC_PERIPH_5V_OC
 	// OC pins are active low
 	system_power.periph_5v_oc  = BOARD_ADC_PERIPH_5V_OC;
-#else
-	system_power.periph_5v_oc  = 0;
 #endif
+
 #ifdef BOARD_ADC_HIPOWER_5V_OC
 	system_power.hipower_5v_oc = BOARD_ADC_HIPOWER_5V_OC;
-#else
-	system_power.hipower_5v_oc  = 0;
 #endif
 
 	/* lazily publish */
-	if (_to_system_power != nullptr) {
-		orb_publish(ORB_ID(system_power), _to_system_power, &system_power);
-
-	} else {
-		_to_system_power = orb_advertise(ORB_ID(system_power), &system_power);
-	}
+	_to_system_power.publish(system_power);
 
 #endif // BOARD_ADC_USB_CONNECTED
 }
 
 uint16_t board_adc_sample(unsigned channel)
 {
```

Comment:
@dagar I think this needs to be re-thought from an atomic point of view.

Reply:
Which part? Are you thinking about the conversion per sample or getting multiple samples (+ system_power) gpios together?